diff --git a/CMakeLists.txt b/CMakeLists.txt
index fab2eb00e..d9d3c2fe0 100644
--- a/cmake/protobuf.cmake
+++ b/cmake/protobuf.cmake
@@ -30,7 +30,7 @@ if(gRPC_PROTOBUF_PROVIDER STREQUAL "module")
 
   if(EXISTS "${PROTOBUF_ROOT_DIR}/CMakeLists.txt")
     set(protobuf_MSVC_STATIC_RUNTIME OFF CACHE BOOL "Link static runtime libraries")
-    add_subdirectory(${PROTOBUF_ROOT_DIR} third_party/protobuf)
+    # add_subdirectory(${PROTOBUF_ROOT_DIR} third_party/protobuf)
     if(TARGET ${_gRPC_PROTOBUF_LIBRARY_NAME})
       set(_gRPC_PROTOBUF_LIBRARIES ${_gRPC_PROTOBUF_LIBRARY_NAME})
     endif()
